R/xml.R

```diff
@@ -0,0 +1,576 @@
+#' @return The argument passed in unmodified in case a user wanted run the
+#' conversion again at a later time.
+#' @author PP March 2017
+#' @export
+run_xml_conversion <- make_run_xml_conversion()
+
+# Note: the methods below explicitly name XML tags as expected by GCAM and/or
+# the model interface headers thus will need to be maintained to be consistent.
+
+#' add_rename_landnode_xml
+#'
+#' Add a table to an XML pipeline that instructs the ModelInterface to rename
+#' LandNodeX to LandNode.  Such a table is necessary to help work around
+#' limitations in the XML processing that node names of the same name can not be
+#' nested with in each other: LandNode/LandNode thus instead we say
+#' LandNode1/LandNode2 and rename as the last step.  Therefore in most cases a
+#' user should add this table near the end of the XML pipeline.
+#' @param dot The current state of the pipeline started from \code{create_xml}
+#' @return A "data structure" to hold the various parts needed to run the model
+#' interface CSV to XML conversion.
+#' @author Pralit Patel
+#' @export
+add_rename_landnode_xml <- function(dot) {
+  land_name_table <- tibble(from=paste0("LandNode", seq(1,5)),to="LandNode")
+
+  add_xml_data(dot, land_name_table, "NodeRename", NULL)
+}
+
+#' add_node_equiv_xml
+#'
+#' Add a table to an XML pipeline that instructs the ModelInterface to treat
+#' tags in the same class as the same when enabled.  Thus not requiring multiple
+#' headers for instance to read in a share-weight into a technology
+#' intermittent-technology or tranTechnology.  A user taking advantage of this
+#' feature would then read this table early in the XML pipeline.
+#' @param dot The current state of the pipeline started from \code{create_xml}
+#' @param equiv_class A name of the equivalence class to add.  This could be any
+#' key in the list \code{XML_NODE_EQUIV}.
+#' @return A "data structure" to hold the various parts needed to run the model
+#' interface CSV to XML conversion.
+#' @author Pralit Patel
+#' @export
+add_node_equiv_xml <- function(dot, equiv_class) {
+  equiv_list <- XML_NODE_EQUIV[[equiv_class]]
+  if(is.null(equiv_list)) {
+    stop("Could not find ", equiv_class, " in XML_NODE_EQUIV")
+  }
+
+  tibble(group.name=equiv_class, col=paste0("tag", seq(1,length(equiv_list))), tag=equiv_list) %>%
+    spread(col, tag) ->
+    equiv_table
+
+  add_xml_data(dot, equiv_table, "EQUIV_TABLE", NULL)
+}
+
+#' generate_level2_data_names
+#'
+#' Generate the list of "level 2 data names" or really a list of column orderings
+#' keyed by the ModelInterface header so that we can ensure tables being sent to be
+#' converted to XML by the ModelInterface have their columns arranged in the order
+#' the ModelInterface is expecting them.
+#' @return A list object where [[header]] <- character vector of column names
+#' corresponding to header.
+#' @author Pralit Patel
+generate_level2_data_names <- function() {
```

Comment:
This shouldn't be a function.  Create the data structure in your workspace and add it to the `data/` directory using `devtools::use_data()`.  This could probably be made an internal data structure.  Don't forget to add a data.R function to document the data structure, as explained in http://r-pkgs.had.co.nz/data.html#documenting-data

Also, see the notes on reproducing the data set in http://r-pkgs.had.co.nz/data.html#data-extdata .  Briefly, this function should go in a source file in 'data-raw/'.  It's not part of the package, but it's still in the repository so that someone maintaining the package (i.e., adding or deleting names) can rerun it and generate a new version of the data.